internal/handlers/common/distinct.go

```diff
@@ -0,0 +1,138 @@
+// Copyright 2021 FerretDB Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package common
+
+import (
+	"fmt"
+
+	"go.uber.org/zap"
+
+	"github.com/FerretDB/FerretDB/internal/types"
+	"github.com/FerretDB/FerretDB/internal/util/lazyerrors"
+)
+
+// DistinctParams contains `distinct` command parameters supported by at least one handler.
+//
+//nolint:vet // for readability
+type DistinctParams struct {
+	DB         string
+	Collection string
+	Key        string
+	Filter     *types.Document
+	Comment    string
+}
+
+// GetDistinctParams returns `distinct` command parameters.
+func GetDistinctParams(document *types.Document, l *zap.Logger) (*DistinctParams, error) {
+	var err error
+
+	unimplementedFields := []string{
+		"collation",
+	}
+	if err = Unimplemented(document, unimplementedFields...); err != nil {
+		return nil, err
+	}
+
+	ignoredFields := []string{
+		"readConcern",
+	}
+	Ignored(document, l, ignoredFields...)
+
+	var dp DistinctParams
+
+	if dp.DB, err = GetRequiredParam[string](document, "$db"); err != nil {
+		return nil, err
+	}
+
+	collectionParam, err := document.Get(document.Command())
+	if err != nil {
+		return nil, err
+	}
+
+	var ok bool
+	if dp.Collection, ok = collectionParam.(string); !ok {
+		return nil, NewCommandErrorMsgWithArgument(
+			ErrInvalidNamespace,
+			fmt.Sprintf("collection name has invalid type %s", AliasFromType(collectionParam)),
+			document.Command(),
+		)
+	}
+
+	if dp.Key, err = GetRequiredParam[string](document, "key"); err != nil {
+		return nil, err
+	}
+
+	if dp.Key == "" {
+		return nil, NewCommandErrorMsg(ErrEmptyFieldPath, "FieldPath cannot be constructed with empty string")
+	}
+
+	if dp.Filter, err = GetOptionalParam(document, "query", dp.Filter); err != nil {
+		return nil, err
+	}
+
+	if dp.Comment, err = GetOptionalParam(document, "comment", dp.Comment); err != nil {
+		return nil, err
+	}
+
+	return &dp, nil
+}
+
+// FilterDistinctValues returns distinct values from the given slice of documents with the given key.
+//
+// If the key is not found in the document, the document is ignored.
+//
+// If the key is found in the document, and the value is an array, each element of the array is added to the result.
+// Otherwise, the value itself is added to the result.
+func FilterDistinctValues(docs []*types.Document, key string) (*types.Array, error) {
+	distinct := types.MakeArray(len(docs))
```

Comment:
I understand that at the end we need to return the array, though why can't we use map here? As the `distinct` command runs against whole collection, iterating through it for every field could not scale well. Map would also remove requirement to sort everything afterwards. Is there any technical difficulty to do that now?

Reply:
Unfortunately we can't use map to check for duplicates because a value of `Binary` type can't be a map key (it's not possible to get a hash from it).

But I like your question!

Reply:
Map will not remove the requirement to sort things because there is no guaranteed order of the keys in the map. And also the result must be of `types.Array`, so if we used map, we had to iterate through it to build an array and sort that array.